### PR TITLE
[opt](fucntion) improve json_extract function handle const column

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes
VARCHAR json_extract(VARCHAR json_str, VARCHAR path[, VARCHAR path] ...)
```
  for (int i = 0; i < arguments.size(); i++) {
       column_ptrs.push_back(
               block.get_by_position(arguments[i]).column->convert_to_full_column_if_const());
       data_columns.push_back(assert_cast<const ColumnString*>(column_ptrs.back().get()));
   }

before not handle const column, as the input arguments are variadic,
and most user case is like: json_extract(column, '$.fparam.nested_2')
so could special deal with two arguments could reuse json document.

```

```
mysql [test]>select count(json_extract(a, '$.fparam.nested_2')) from json_table_2;
+---------------------------------------------+
| count(json_extract(a, '$.fparam.nested_2')) |
+---------------------------------------------+
|                                    10000001 |
+---------------------------------------------+
1 row in set (1.06 sec)

mysql [test]>select count(json_extract(a, '$.fparam.nested_2')) from json_table_2;
+---------------------------------------------+
| count(json_extract(a, '$.fparam.nested_2')) |
+---------------------------------------------+
|                                    10000001 |
+---------------------------------------------+
1 row in set (1.02 sec)

mysql [test]>
mysql [test]>
mysql [test]>select count(json_extract(a, '$.fparam.nested_2')) from json_table_2;
+---------------------------------------------+
| count(json_extract(a, '$.fparam.nested_2')) |
+---------------------------------------------+
|                                    10000001 |
+---------------------------------------------+
1 row in set (44.22 sec)

mysql [test]>select count(json_extract(a, '$.fparam.nested_2')) from json_table_2;
+---------------------------------------------+
| count(json_extract(a, '$.fparam.nested_2')) |
+---------------------------------------------+
|                                    10000001 |
+---------------------------------------------+
1 row in set (42.80 sec)

```

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/functions/function_json.cpp [hunk 1]
```diff
@@ -826,44 +826,60 @@ struct FunctionJsonExtractImpl {
...
     static rapidjson::Value parse_json(const ColumnString* json_col, const ColumnString* path_col,
-                                       rapidjson::Document::AllocatorType& allocator,
-                                       const int row) {
+                                       rapidjson::Document::AllocatorType& allocator, const int row,
+                                       const int col, std::vector<bool>& column_is_consts) {
         rapidjson::Value value;
         rapidjson::Document document;
 
-        const auto obj = json_col->get_data_at(row);
+        const auto& obj = json_col->get_data_at(index_check_const(row, column_is_consts[0]));
```
Comment (HappenLee) [inline be/src/vec/functions/function_json.cpp:834]: here is `StringRef`, seem do not need use & as ref

File: be/src/vec/functions/function_json.cpp [hunk 2]
```diff
@@ -826,44 +826,60 @@ struct FunctionJsonExtractImpl {
...
             value.CopyFrom(*root, allocator);
         }
         return value;
     }
 
+    static rapidjson::Value* get_document(const ColumnString* path_col,
+                                          rapidjson::Document* document,
+                                          std::vector<JsonPath>& parsed_paths, const int row,
+                                          bool is_const_column) {
+        const auto& path = path_col->get_data_at(index_check_const(row, is_const_column));
```
Comment (HappenLee) [inline be/src/vec/functions/function_json.cpp:849]: same to up

File: be/src/vec/functions/function_json.cpp [hunk 3]
```diff
@@ -826,44 +826,60 @@ struct FunctionJsonExtractImpl {
...
+            return nullptr;
+        }
+
+#ifdef USE_LIBCPP
+        std::string s(path_string);
+        auto tok = get_json_token(s);
+#else
+        auto tok = get_json_token(path_string);
+#endif
+        std::vector<std::string> paths(tok.begin(), tok.end());
```
Comment (HappenLee) [inline be/src/vec/functions/function_json.cpp:863]: Add todo here is there can be `std::vector<std::string>` or `std::span` ？

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by zhangstar333** [comment]
run buildall

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**